plugin/kubernetes/xfr.go

```diff
@@ -1,17 +1,92 @@
 package kubernetes
 
 import (
-	"time"
+	"net"
+	"strings"
 
+	"github.com/coredns/coredns/plugin/etcd/msg"
 	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+	api "k8s.io/api/core/v1"
 )
 
 // Serial implements the Transferer interface.
-func (e *Kubernetes) Serial(state request.Request) uint32 {
-	return uint32(time.Now().Unix())
-}
+func (k *Kubernetes) Serial(state request.Request) uint32 { return uint32(k.APIConn.Modified()) }
 
 // MinTTL implements the Transferer interface.
-func (e *Kubernetes) MinTTL(state request.Request) uint32 {
-	return 30
+func (k *Kubernetes) MinTTL(state request.Request) uint32 { return 30 }
+
+// Transfer implements the Transferer interface.
+func (k *Kubernetes) Transfer(state request.Request) <-chan msg.Service {
+	c := make(chan msg.Service)
+
+	go k.transfer(c, state.Zone)
+
+	return c
+}
+
+func (k *Kubernetes) transfer(c chan msg.Service, zone string) {
+
+	defer close(c)
+
+	zonePath := msg.Path(zone, "coredns")
+	serviceList := k.APIConn.ServiceList()
+	for _, svc := range serviceList {
+		switch svc.Spec.Type {
+		case api.ServiceTypeClusterIP, api.ServiceTypeNodePort, api.ServiceTypeLoadBalancer:
+
+			if net.ParseIP(svc.Spec.ClusterIP) != nil {
+				for _, p := range svc.Spec.Ports {
+
+					s := msg.Service{Host: svc.Spec.ClusterIP, Port: int(p.Port), TTL: k.ttl}
+					s.Key = strings.Join([]string{zonePath, Svc, svc.Namespace, svc.Name}, "/")
```

Comment:
I think we should return RRs here if possible instead of msg.Services, you also don't need `HostType` because you make that decision here.

This also means the transfer interface need to be changed, just like it did in my PR

Reply:
The one in #1325 ?